pkg/storegateway/bucket_index_postings.go

```diff
@@ -322,3 +324,143 @@ func checkNilPosting(l labels.Label, p index.Postings) index.Postings {
 	}
 	return p
 }
+
+type postingsSelectionStrategy interface {
+	// name should be a static string which identifies this strategy.
+	// The return value can be later used to find the strategy which yielded a partitioning of postingGroups.
+	name() string
+
+	// selectPostings can modify the passed slice of posting groups.
+	selectPostings([]postingGroup) (selected, omitted []postingGroup)
+}
+
+type selectAllStrategy struct{}
+
+func (selectAllStrategy) name() string {
+	return "all"
+}
+
+func (selectAllStrategy) selectPostings(groups []postingGroup) (selected, omitted []postingGroup) {
+	return groups, nil
+}
+
+// worstCaseFetchedDataStrategy select a few of the posting groups such that their total size
+// does not exceed the size of series in the worst case. The worst case is fetching all series
+// in the smallest non-subtractive posting group - this is effectively the
+// upper bound on how many series all the posting groups can select after being intersected/subtracted.
+//
+// This strategy is meant to prevent fetching some the largest posting lists in an index. Those are usually
+// less selective the smaller ones and only add cost to fetching.
+//
+// The strategy greedily selects the first N posting groups (sorted in ascending order by their size)
+// whose combined size doesn't exceed the size of series in the worst case. The rest of the posting groups
+// are omitted.
+// worstCaseFetchedDataStrategy uses a fixed estimation about the size of series in the index (tsdb.EstimatedSeriesP99Size).
+type worstCaseFetchedDataStrategy struct {
+	// postingListActualSizeFactor affects how
+	// posting lists are summed together.
+	// This factor is multiplied by the size of posting lists
+	// in the block index.
+	postingListActualSizeFactor float64
+}
+
+func (s worstCaseFetchedDataStrategy) name() string {
+	return fmt.Sprintf("worstCase%0.1f", s.postingListActualSizeFactor)
+}
+
+func (s worstCaseFetchedDataStrategy) selectPostings(groups []postingGroup) (selected, omitted []postingGroup) {
+	const seriesBytesPerPostingByte = tsdb.EstimatedSeriesP99Size / tsdb.BytesPerPostingInAPostingList
+
+	sort.Slice(groups, func(i, j int) bool {
+		return groups[i].totalSize < groups[j].totalSize
+	})
+
+	var minGroupSize int64
+	for _, g := range groups {
+		// The size of each posting list contains 4 bytes with the number of entries.
+		// We shouldn't count these as series.
+		if !g.isSubtract && !(len(g.keys) == 1 && g.keys[0] == allPostingsKey) {
+			minGroupSize = g.totalSize - int64(len(g.keys)*4)
```

Comment:
Question: should we apply `s.postingListActualSizeFactor` here as well? If that's the factor we have to apply to compute the actual estimated group size, then why not applying it here too?

Reply:
The min group size is used to calculate the number of series. The number of series is the same whether the list is fetched from the cache or the bucket. This wasn't very clear in the way the code was written. I made a bit more explicit in db2cf160056f6d8b7af4ec71969be127dcecbc0a. PTAL